File: sdks/python/kimberlite/tenant_pool.py
"""Per-tenant :class:`Client` cache.

AUDIT-2026-04 S2.4 — lifts notebar's LRU-per-tenant adapter out
of ``packages/kimberlite-client/src/adapter.ts`` into the SDK so
every multi-tenant Python app (FastAPI, Django, Flask) gets the
same pattern.

Typical FastAPI use — one pool per process, ``.acquire(tenant_id)``
in each handler::

    from kimberlite import Client
    from kimberlite.tenant_pool import TenantPool

    pool = TenantPool(
        factory=lambda tid: Client.connect(
            addresses=["localhost:5432"], tenant_id=tid, auth_token=TOKEN,
        ),
        max_size=128,
        idle_timeout_ms=5 * 60_000,
    )

    @app.get("/charts/{patient_id}")
    def chart(patient_id: int, tenant_id: int = Depends(current_tenant)):
        client = pool.acquire(tenant_id)
        return client.query("SELECT ...")
"""

from __future__ import annotations

import threading
import time
from collections import OrderedDict
from dataclasses import dataclass
from typing import Callable, Optional

from .client import Client


@dataclass(frozen=True)
class TenantPoolStats:
    """Runtime stats snapshot. Useful for dashboards + tests."""

    size: int
    hits: int
    misses: int
    evictions: int
    idle_evictions: int


class TenantPool:
    """LRU-per-tenant :class:`Client` cache.

    Thread-safe: all mutations hold an internal lock so concurrent
    FastAPI handlers can call :meth:`acquire` without racing.
    Concurrent acquires for the same ``tenant_id`` deduplicate —
    only one ``factory()`` call fires.
    """

    def __init__(
        self,
        factory: Callable[[int], Client],
        max_size: int = 128,
        idle_timeout_ms: int = 5 * 60_000,
        *,
        now_ms: Callable[[], int] = lambda: int(time.time() * 1000),
    ) -> None:
        """Create a new pool.

        Args:
            factory: Callable that opens a fresh :class:`Client`
                bound to the given ``tenant_id``. Typically
                ``lambda tid: Client.connect(addresses, tid, ...)``.
            max_size: Maximum concurrent cached tenants. LRU-evicted
                above this.
            idle_timeout_ms: Idle-timeout in milliseconds. Clients
                untouched for this long are disconnected + evicted on
                the next :meth:`acquire` call. ``0`` disables idle
                eviction.
            now_ms: Injectable clock (for deterministic tests). Returns
                milliseconds.
        """
        self._factory = factory
        self._max_size = max_size
        self._idle_timeout_ms = idle_timeout_ms
        self._now_ms = now_ms
        # OrderedDict preserves insertion order + O(1) move-to-end
        # for LRU tracking.
        self._cache: "OrderedDict[int, tuple[Client, int]]" = OrderedDict()
        self._lock = threading.RLock()
        self._inflight: dict[int, threading.Event] = {}
        self._hits = 0
        self._misses = 0
        self._evictions = 0
        self._idle_evictions = 0

    def acquire(self, tenant_id: int) -> Client:
        """Return the cached :class:`Client` for ``tenant_id``,
        creating one via the factory if absent.

        Updates the LRU recency stamp. Concurrent calls for the same
        ``tenant_id`` block until the first acquirer finishes
        connecting.
        """
        with self._lock:
            self._expire_idle()
            entry = self._cache.get(tenant_id)
            if entry is not None:
                client, _ = entry
                now = self._now_ms()
                self._cache.move_to_end(tenant_id)
                self._cache[tenant_id] = (client, now)
                self._hits += 1
                return client

            # Inflight dedup: if another caller is already
            # connecting for this tenant, wait for them.
            inflight_event = self._inflight.get(tenant_id)
            if inflight_event is not None:
                self._hits += 1
                # Release lock while waiting — the other caller
                # holds it to insert the result.
                self._lock.release()
                try:
                    inflight_event.wait()
                finally:
                    self._lock.acquire()
                entry = self._cache.get(tenant_id)
                if entry is not None:
                    self._cache.move_to_end(tenant_id)
                    return entry[0]
                # Fall through to re-create if the inflight acquirer
                # failed. Rare — typically the inflight insert
                # succeeds and we return above.

            self._misses += 1
            event = threading.Event()
            self._inflight[tenant_id] = event

        # Drop the lock while the factory runs — it may do network
        # I/O. The inflight event guarantees dedup.
        try:
            client = self._factory(tenant_id)
        except Exception:
            with self._lock:
                self._inflight.pop(tenant_id, None)
            event.set()
            raise

        with self._lock:
            self._evict_if_full()
            self._cache[tenant_id] = (client, self._now_ms())
            self._inflight.pop(tenant_id, None)
        event.set()
        return client

    def with_client(
        self, tenant_id: int, fn: Callable[[Client], object]
    ) -> object:
        """Convenience — :meth:`acquire` + execute. The client is
        never surfaced to the caller past ``fn``'s scope.
        """
        client = self.acquire(tenant_id)
        return fn(client)

    def close(self) -> None:
        """Drop all cached clients, disconnecting each one.

        Subsequent :meth:`acquire` calls reconnect via the factory.
        """
        with self._lock:
            to_close = list(self._cache.values())
            self._cache.clear()
        for client, _ in to_close:
            try:
                client.disconnect()
            except Exception:  # pragma: no cover — best-effort
                pass

    def stats(self) -> TenantPoolStats:
        """Runtime stats snapshot."""
        with self._lock:
            return TenantPoolStats(
                size=len(self._cache),
                hits=self._hits,
                misses=self._misses,
                evictions=self._evictions,
                idle_evictions=self._idle_evictions,
            )

    # ------------------------------------------------------------------

    def _evict_if_full(self) -> None:
        """Evict the least-recently-used entry when the cache is
        at or above ``max_size``. Caller must hold ``self._lock``.
        """
        while len(self._cache) >= self._max_size:
            # popitem(last=False) removes the OLDEST entry —
            # head of the ordered dict.
            oldest_id, (client, _) = self._cache.popitem(last=False)
            try:
                client.disconnect()
            except Exception:  # pragma: no cover
                pass
            self._evictions += 1

    def _expire_idle(self) -> None:
        """Evict entries whose ``last_used_at`` is older than the
        idle cutoff. Caller must hold ``self._lock``.
        """
        if self._idle_timeout_ms == 0:
            return
        cutoff = self._now_ms() - self._idle_timeout_ms
        stale = [
            tid for tid, (_, last) in self._cache.items() if last < cutoff
        ]
        for tid in stale:
            client, _ = self._cache.pop(tid)
            try:
                client.disconnect()
            except Exception:  # pragma: no cover
                pass
            self._idle_evictions += 1

File: sdks/typescript/src/tenant-pool.ts
/**
 * Per-tenant Client cache.
 *
 * AUDIT-2026-04 S2.4 — lifts notebar's LRU-per-tenant adapter out
 * of `packages/kimberlite-client/src/adapter.ts` into the SDK so
 * every multi-tenant SaaS gets the same pattern without
 * re-implementing LRU + idle eviction + factory callbacks.
 *
 * Difference from `Pool`:
 * - `Pool` multiplexes N connections to a SINGLE tenant across
 *   concurrent callers.
 * - `TenantPool` holds one long-lived `Client` per tenantId so
 *   N concurrent tenants can talk to the server without
 *   reconnecting on every call.
 *
 * Typical React Router V7 / Express use — one pool per process,
 * `.acquire(tenantId)` in each handler:
 *
 * ```ts
 * const pool = new TenantPool({
 *   factory: (tenantId) => Client.connect({ addresses: [addr], tenantId }),
 *   maxSize: 128,
 *   idleTimeoutMs: 5 * 60_000,
 * });
 *
 * export async function loader({ params }) {
 *   const client = await pool.acquire(BigInt(params.tenantId));
 *   return json(await client.query('...'));
 * }
 * ```
 */

import { Client } from './client';

type TenantId = bigint;

/**
 * Configuration for {@link TenantPool}.
 */
export interface TenantPoolConfig {
  /**
   * Callback that opens a fresh `Client` for a tenant. Typically
   * `(id) => Client.connect({ addresses, tenantId: id, authToken })`.
   */
  readonly factory: (tenantId: TenantId) => Promise<Client>;
  /**
   * Maximum number of concurrent cached tenants. When exceeded,
   * the least-recently-used client is disconnected + evicted.
   * Defaults to 128.
   */
  readonly maxSize?: number;
  /**
   * Idle-timeout in milliseconds. Clients untouched for this long
   * are disconnected + evicted on the next `acquire()` call.
   * `0` disables idle eviction. Defaults to 5 minutes.
   */
  readonly idleTimeoutMs?: number;
  /**
   * AUDIT-2026-04 S4.10 — callback invoked when a cached client
   * reports an authentication failure. Typically re-mints a
   * short-lived JWT from a long-lived refresh token. The pool
   * evicts the stale client, then calls {@link factory} again to
   * open a fresh connection with the new token.
   *
   * Without this callback, rotating service-account tokens requires
   * rebuilding the pool from scratch — notebar's feedback was that
   * this is painful in prod because it kicks every cached tenant.
   *
   * Return `null` to opt out (pool re-raises the auth failure).
   */
  readonly refreshToken?: (tenantId: TenantId) => Promise<string | null>;
  /**
   * AUDIT-2026-04 S4.11 — optional push callback invoked on every
   * internal pool event (hit / miss / eviction / idle-eviction /
   * refresh). Apps typically wire an OTel counter or metrics
   * dashboard exporter to this so pool pressure is visible without
   * polling {@link stats}.
   */
  readonly onStats?: (event: TenantPoolEvent, stats: TenantPoolStats) => void;
  /**
   * Injectable clock for deterministic tests. Defaults to
   * `Date.now`.
   */
  readonly now?: () => number;
}

/** Event tags for {@link TenantPoolConfig.onStats}. */
export type TenantPoolEvent =
  | 'hit'
  | 'miss'
  | 'eviction'
  | 'idle-eviction'
  | 'refresh';

/**
 * Runtime statistics — exposed for dashboards and tests.
 */
export interface TenantPoolStats {
  readonly size: number;
  readonly hits: number;
  readonly misses: number;
  readonly evictions: number;
  readonly idleEvictions: number;
}

interface Entry {
  client: Client;
  lastUsedAt: number;
}

/**
 * LRU-per-tenant `Client` cache.
 */
export class TenantPool {
  private readonly clients = new Map<TenantId, Entry>();
  private readonly factory: (tenantId: TenantId) => Promise<Client>;
  private readonly maxSize: number;
  private readonly idleTimeoutMs: number;
  private readonly refreshToken?: (tenantId: TenantId) => Promise<string | null>;
  private readonly onStats?: (event: TenantPoolEvent, stats: TenantPoolStats) => void;
  private readonly now: () => number;

  private hits = 0;
  private misses = 0;
  private evictions = 0;
  private idleEvictions = 0;
  // Per-tenant connection promises — dedup concurrent acquires
  // for the same tenantId so only one `factory()` call fires.
  private readonly inflight = new Map<TenantId, Promise<Client>>();

  constructor(cfg: TenantPoolConfig) {
    this.factory = cfg.factory;
    this.maxSize = cfg.maxSize ?? 128;
    this.idleTimeoutMs = cfg.idleTimeoutMs ?? 5 * 60_000;
    this.refreshToken = cfg.refreshToken;
    this.onStats = cfg.onStats;
    this.now = cfg.now ?? Date.now;
  }

  /**
   * AUDIT-2026-04 S4.10 — evict the cached client for `tenantId`
   * (if any), call {@link TenantPoolConfig.refreshToken} to mint a
   * fresh credential, then reconnect via {@link factory}. Returns
   * the new client. Consumers typically call this from an
   * error-handler when they observe an `AuthenticationFailed`
   * error on a long-lived pool entry.
   *
   * Throws if no `refreshToken` callback was configured or if it
   * returned `null`.
   */
  async refresh(tenantId: TenantId): Promise<Client> {
    if (this.refreshToken === undefined) {
      throw new Error(
        `TenantPool.refresh(${tenantId}): refreshToken callback is not configured`,
      );
    }
    const existing = this.clients.get(tenantId);
    if (existing !== undefined) {
      this.clients.delete(tenantId);
      void existing.client.disconnect().catch(() => undefined);
    }
    const token = await this.refreshToken(tenantId);
    if (token === null) {
      throw new Error(
        `TenantPool.refresh(${tenantId}): refreshToken returned null`,
      );
    }
    // The factory closes over the app's Client.connect config; the
    // refreshed token is expected to have been applied at that
    // level (e.g. by stashing it on a mutable slot the factory
    // reads). The pool just forces the reconnect.
    const client = await this.factory(tenantId);
    this.clients.set(tenantId, { client, lastUsedAt: this.now() });
    this.emit('refresh');
    return client;
  }

  private emit(event: TenantPoolEvent): void {
    if (this.onStats !== undefined) {
      this.onStats(event, this.stats());
    }
  }

  /**
   * Return the cached `Client` for `tenantId`, creating one via
   * the factory if absent. Updates the LRU recency stamp.
   *
   * Concurrent calls for the same `tenantId` deduplicate — only
   * one `factory()` call fires.
   */
  async acquire(tenantId: TenantId): Promise<Client> {
    this.expireIdle();

    const entry = this.clients.get(tenantId);
    if (entry !== undefined) {
      entry.lastUsedAt = this.now();
      this.hits += 1;
      this.emit('hit');
      return entry.client;
    }

    // Dedup concurrent misses for the same tenant.
    const pending = this.inflight.get(tenantId);
    if (pending !== undefined) {
      this.hits += 1;
      this.emit('hit');
      return pending;
    }

    this.misses += 1;
    const p = this.connectAndInsert(tenantId);
    this.inflight.set(tenantId, p);
    try {
      const c = await p;
      this.emit('miss');
      return c;
    } finally {
      this.inflight.delete(tenantId);
    }
  }

  /**
   * Convenience — `acquire` + execute + updates lastUsedAt. The
   * client itself is never surfaced; useful when callers should
   * not retain a handle past the call's scope.
   */
  async withClient<T>(
    tenantId: TenantId,
    fn: (client: Client) => Promise<T>,
  ): Promise<T> {
    const client = await this.acquire(tenantId);
    return fn(client);
  }

  /**
   * Drop all cached clients, disconnecting each one. Subsequent
   * `acquire()` calls will reconnect via the factory.
   */
  async close(): Promise<void> {
    const toClose = Array.from(this.clients.values());
    this.clients.clear();
    await Promise.all(toClose.map((e) => e.client.disconnect().catch(() => undefined)));
  }

  /** Runtime stats snapshot. */
  stats(): TenantPoolStats {
    return {
      size: this.clients.size,
      hits: this.hits,
      misses: this.misses,
      evictions: this.evictions,
      idleEvictions: this.idleEvictions,
    };
  }

  private async connectAndInsert(tenantId: TenantId): Promise<Client> {
    this.evictIfFull();
    const client = await this.factory(tenantId);
    this.clients.set(tenantId, { client, lastUsedAt: this.now() });
    return client;
  }

  private evictIfFull(): void {
    if (this.clients.size < this.maxSize) return;
    let oldestKey: TenantId | null = null;
    let oldestAt = Number.POSITIVE_INFINITY;
    for (const [key, entry] of this.clients.entries()) {
      if (entry.lastUsedAt < oldestAt) {
        oldestAt = entry.lastUsedAt;
        oldestKey = key;
      }
    }
    if (oldestKey !== null) {
      const e = this.clients.get(oldestKey);
      if (e !== undefined) {
        void e.client.disconnect().catch(() => undefined);
      }
      this.clients.delete(oldestKey);
      this.evictions += 1;
      this.emit('eviction');
    }
  }

  private expireIdle(): void {
    if (this.idleTimeoutMs === 0) return;
    const cutoff = this.now() - this.idleTimeoutMs;
    const stale: TenantId[] = [];
    for (const [key, entry] of this.clients.entries()) {
      if (entry.lastUsedAt < cutoff) stale.push(key);
    }
    for (const key of stale) {
      const e = this.clients.get(key);
      if (e !== undefined) {
        void e.client.disconnect().catch(() => undefined);
      }
      this.clients.delete(key);
      this.idleEvictions += 1;
      this.emit('idle-eviction');
    }
  }
}
